### PR TITLE
[react-native] #46426 types of property key are incompatible

### LINKED ISSUE #46426: Types of property 'key' are incompatible. All animated components props are wrapper with `WithAnimatedValue`
Is something wrong with CI or publishing infrastructure?

- Check the [Infrastructure status updates](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/44317) thread for known issues.
- Leave a message on the [DefinitelyTyped Gitter](http://gitter.im/DefinitelyTyped/DefinitelyTyped).

If you know how to fix the issue, make a pull request instead.

- [x] I tried using the `@types/react-native` package and had problems.
- [ ] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [ ] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @alloy @huhuanming @iRoachie 

If you do not mention the authors the issue will be ignored.

At the moment `withAnimatedValue` adds `Value | AnimatedInterpolation` to all props of `Animated` components. On top of that, it's also extending interface which means primitive types are not handled properly:
```
 type Test = number | string extends WithAnimatedValue<number | string> ? true : false // is false
```

Given that it causes some issues for code like:

```
type Props = { type: "brand" } & React.ComponentPropsWithoutRef<typeof Animated.Text>;

const TextWrapper: React.FunctionComponent<Props> = ({ brand, ...props  })=> <Animated.Text {...props} />
```

```
  Types of property 'key' are incompatible.         
      Type 'WithAnimatedValue<string | number | undefined> | undefined' is not assignable to type 'string | number | undefined'.           **Type 'WithAnimatedValue<string | number | undefined>' is not assignable to type 'string | number | undefined'**.             Type 'WithAnimatedValue<string | number | undefined>' is not assignable to type 'number'.``
```

We should refactor `WithAnimatedValue` to be only applied to props that really can handle `Value | AniimateInterpolation` type.

Below is the base example that will make it work properly but that it still needs some polishing and reviewing whether there is any kind of animated props that can also handle `Value` except `style` prop.  
```
    export type WithAnimatedValue<T> = StyleProp<T extends StyleProp<infer R> ? (R extends object
                ? { [K in keyof R]?: WithAnimatedValue<R[K]> }
                : R extends (infer P)[]
                ? WithAnimatedValue<P>[]
                : R | Value | AnimatedInterpolation) : never>

    export interface AnimatedComponent<T extends React.ComponentType<any>>
        extends React.FC<Omit<ComponentProps<T>, "style"> & { style?: WithAnimatedValue<ComponentProps<T>["style"]> }> {}
````

### PR DESCRIPTION
**Pull request by desfero**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: fixes #46426
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

File: types/react-native/index.d.ts [hunk 1]
```diff
@@ -8939,24 +8939,39 @@ export namespace Animated {
...
+        ? T
+        : T extends Primitive // add `Value` and `AnimatedInterpolation` but also preserve original T
+        ? T | Value | AnimatedInterpolation
+        : T extends Array<infer P>
+        ? WithAnimatedArray<P>
+        : T extends {}
+        ? WithAnimatedObject<T> // in case it's something we don't yet know about (for .e.g bigint)
+        : T;
+
+    type NonAnimatedProps = 'key' | 'ref';
```
Comment (desfero) [inline types/react-native/index.d.ts:8961]: I'm not really sure we should augment all props with `Value and AnimatedInterpolation`. 
I tried to find any kind of useful information in docs or react-native source code where `Animated.Value` and `AnimatedInterpolation` can be passed as a prop (except `style` prop) to narrow augmentation but I just don't see any kind of list of supported by animated values props.

**Review by SerhiyZheliznjak — APPROVED** [review]

**Comment by desfero** [comment]
Ready to merge

**Comment by alloy** [comment]
@desfero I have a bit of a hard time following the exact issue and the fix, also your inline comment. Can you expand on what the status is of these changes, especially seeing as you also made another change that already landed?

**Comment by alloy** [comment]
@orta In this PR the OP was able to self-merge changes to the RN DT package, because a maintainer of _another_ DT package approved those other changes. I wonder if you would consider this a bug?

**Comment by desfero** [comment]
@alloy sorry for the late reply. 
So the problem was in `AnimatedProps` as it augments all component Animated components props with `Animated.Value` as it's not valid given for .e.g. `key` prop does not expect an animated value. That in turn makes all `React.ComponentProps` to infer incompatible types as `key`s are no longer assignable to each other given the augmented one also allows `Animiated.Value`. I'm not fully happy with the approach here as i would even prefer to narrow augmentation to only props that truly can handle `Animated.Value`s. 
This PR also introduced a regression bug I didn't catch so the next one is just a fix with a unit test to be covered.

@alloy @orta I do also agree that it should not be possible to merge PR in case the maintainer of another package approved PR changes.

**Comment by alloy** [comment]
@desfero No worries, thanks for taking the time to elaborate! 🙏 

I’ll move the report about the ability to self-merge to the appropriate repo.

**Pull request merged**